torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -907,6 +908,31 @@ bool ProcessGroupNCCL::CoalescedWorkNCCL::wait(
   return true;
 }
 
+ProcessGroupNCCL::DebugInfoCallbackStorer::DebugInfoCallbackStorer(
+    std::string fileName)
+    : fileName_(std::move(fileName)) {}
+
+ProcessGroupNCCL::DebugInfoCallbackStorer::~DebugInfoCallbackStorer() = default;
+
+void ProcessGroupNCCL::DebugInfoCallbackStorer::storeTraceInfoStorer(
+    int rank,
+    const std::string& ncclTrace) {
+  auto filename = c10::str(std::string(fileName_), rank);
+
+  // Open a file for writing. The ios::binary flag is used to write data as
+  // binary.
+  std::ofstream file(filename, std::ios::binary);
+
+  // Check if the file was opened successfully.
+  if (!file.is_open()) {
+    LOG(ERROR) << "Error opening file for writing NCCLPG debug info: "
+               << filename;
+    return;
+  }
+
+  file.write(ncclTrace.data(), ncclTrace.size());
```

Comment:
wonder if we're gonna need the same solution for flushing this file, as we need for the test case.

in fact, i wonder if you even need to subclass the writer.  can you just open the tempfile inside the test case and then pass the path to the tempfile to the constructor for this writer, and fully reuse it?

Reply:
Actually since we are using streaming and it will automatically flush one the function exit. Unfortunately, `std::tmpfile` does not support a given path... Read/write using stream is much better than using fgets/fputs, but looks like std::tmpfile does not work with stream either.. One alternative here is that we manually manage the file removal ourselves so that we can generate a temporary file name using `std::tmpnam` and use std::remove in the end to remove the file. WDYT?

Reply:
Actually, from cpp 17, there is something called std::filesystem; we can use it to replace `std::tmpfile`. But in the end, we still need to do something like: `fs::remove(tempPath);`

Reply:
OK, I am able to just add a slight change to the write and reuse most of logic from the default writer. Also we can still use the path set from Env variable by using `std::filesystem::temp_directory_path` and `std::filesystem::remove`.